Add support for Intel RDT

marquiz opened the pull request:
OCI runtime-spec has supported Intel RDT (resctrl pseudo-filesystem) for a while already. This RFC PR adds two independent parts of RDT support to containerd.

The first one is integration with github.com/intel/goresctrl by adding a new internal `rdt` plugin. This enables flexible class-based configuration mechanism for RDT (via resctrl fs). The concept is simple: in containerd level only specify a path to an external configuration file. The structure of the config file and the logic of applying it is owned by the outside library. This integration will give users an easy option for configuring RDT in tandem with the container runtime.

The second part adds support for a CRI-namespaced container annotation for controlling the RDT class/CLOS. The idea is that Kubelet can utilize this annotation before the CRI API properly supports RDT. There's also a patch that adds respective Pod annotation for controlling the RDT class (per-pod and per-container) – this is for testing/demonstration purposes of K8s integration.

My take is that the next steps – e.g. how to split this PR, how to proceed with CRI API etc – depend on how this PR is received.

Related, required runc PR:
https://github.com/opencontainers/runc/pull/2920

Corresponding PR submitted against CRI-O:
https://github.com/cri-o/cri-o/pull/4830

askervin commented on pkg/rdt/rdt.go:
```diff
@@ -0,0 +1,72 @@
+	"github.com/containerd/containerd/log"
+	"github.com/intel/goresctrl/pkg/rdt"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/yaml"
```
sigs.k8s.io/yaml not used.

marquiz replied:
Fixed

marquiz commented:
> Can we make it compile-time optional ? Obviously this only makes sense on (recent enough) Intel machines.

It starts to be quite ubiquitous, I'd say, as it's been around since 2015 or smth. We could make it optional, sure, but I'd have it on by default (i.e. a build tag to disable it).

samuelkarp commented:
/ok-to-test

(One of the committers will need to hit the button for the GitHub Actions workflows; I can't do that)

> It starts to be quite ubiquitous, I'd say, as it's been around since 2015 or smth. We could make it optional, sure, but I'd have it on by default (i.e. a build tag to disable it).

Does this work on non-Intel x86_64 chips or on other architectures?

marquiz commented:
> Does this work on non-Intel x86_64 chips or on other architectures?

AMD chips support this (they call it platform qos or smth). Other architectures, I don't think so

marquiz commented:
> Other architectures, I don't think so

Actually, I think I was wrong:
https://lkml.org/lkml/2020/2/14/2451

marquiz commented on docs/man/containerd-config.toml.5.md:
```diff
@@ -83,6 +83,7 @@ documentation.
    **mutation_threshold** guarantees GC is scheduled after n number of database mutations (Default: **100**),
    **schedule_delay** defines the delay after trigger event before scheduling a GC (Default **"0ms"** [immediate]),
    **startup_delay** defines the delay after startup before scheduling a GC (Default **"100ms"**)
+ - **[plugins.rdt]** has one option __rdt_config_file__ (Default: **""**)
```
Good catch, fixed.

marquiz commented on pkg/rdt/rdt.go:
```diff
@@ -0,0 +1,72 @@
+package rdt
+
+import (
+	"io/ioutil"
```
Fixed

marquiz commented:
Rebased and updated. @askervin's review comments addressed. This PR is now more or less aligned with the corresponding PR against CRI-O (cri-o/cri-o#4830).

There were some changes that I wanted to make wrt to the initial version of the patchset. 

One is moving the annotation parsing to the goresctrl library. Main motivation of this change is code re-usability between container runtimes. Behind the scenes this change enables validation of class names (non-existing classes are rejected). It also adds (optional) access control for classes, making it possible to specify certain classes to be unavailable via container and/or pod annotations.

Another smaller addition is storing enabled/disabled status inside the RDT plugin. I wasn't sure what would be a preferred way to store the internal state of a plugin so I just used a global variable (unexported). The status is used to determine whether to act on the specified RDT class or not. In the current implementation the RDT class is simply ignored if RDT plugin is in disabled state. I'm just pondering if we should return an error instead 🤔 WDYT?

samuelkarp commented:
> The status is used to determine whether to act on the specified RDT class or not. In the current implementation the RDT class is simply ignored if RDT plugin is in disabled state. I'm just pondering if we should return an error instead :thinking: WDYT?

If RDT support is requested for a container but can't be set because the plugin is disabled, that seems like it should be an error to me.

marquiz commented:
Rebased

> If RDT support is requested for a container but can't be set because the plugin is disabled, that seems like it should be an error to me.

Yeah, I think I'm inclined to agree. Changed the code as you suggested.

marquiz commented:
Removed RFC status from the title of this PR

marquiz commented:
Rebased

@metux what do you think about the build tag for disabling this, do you think it's needed (I haven't added that, yet)?

marquiz commented:
/retest

marquiz commented:
ping @AkihiroSuda @crosbymichael @dmcgowan @estesp @mlaventure @stevvooe @Random-Liu @mikebrow @fuweid @mxpv @dims @kevpar

Any feedback/comments?

mxpv commented on go.mod:
```diff
@@ -44,27 +45,26 @@ require (
 	gotest.tools/v3 v3.0.3
 	k8s.io/api v0.20.6
-	k8s.io/apimachinery v0.20.6
+	k8s.io/apimachinery v0.21.2
```
Are these vendor updates relevant to rdt?

marquiz replied:
Go mod does that automatically when doing `go get github.com/intel/goresctrl`

`goresctrl` depends on k8s.io because we utilize `k8s.io/apimachinery/pkg/api/resource.Quantity` in some parts of the library.

mxpv commented on pkg/rdt/rdt.go:
```diff
@@ -0,0 +1,90 @@
+
+	enabled = true
+
+	return nil, nil
```
I'm not sure if we should use plugins this way.
Usually when we do
```golang
i, err := p.Instance()
if err != nil {
    return nil, err
}
```
we expect `i` to be a valid object that is safe to use.

Would `sync.Once` be a better fit here?

mxpv replied:
So we try to initialize on first invoke of `ContainerClassFromAnnotations`

samuelkarp replied:
Either `sync.Once` or `sync.RWMutex` around `enabled` seems like a better approach.

marquiz replied:
> I'm not sure if we should use plugins this way.
> Usually when we do
> 
> ```go
> i, err := p.Instance()
> if err != nil {
>     return nil, err
> }
> ```
> 
> we expect `i` to be a valid object that is safe to use.
> 
> Would `sync.Once` be a better fit here?

The reason I implemented this as a plugin was that it provided a nice integration point, both in terms of configuration and initialization at runtime startup time.

What I would like to have is to bail out at (runtime) startup phase in case of invalid configuration, for example. And not suddenly error out at the first occasion somebody tries to assign a container to RDT. How is the object returned by InitFn supposed to be used? Is there a certain interface that `InitFn` of internal plugins is expected to implement or some other way to achieve this? How about returning an empty struct?

Any advice or thoughts?

marquiz replied:
Any comments @mxpv @samuelkarp ?

marquiz replied:
Woof @mxpv @samuelkarp! Any thoughts on that^? The corresponding PR in CRI-O (cri-o/cri-o#4830) was merged a while ago so I'd like to get containerd on par with that 😉

mikebrow replied:
since this seems to be a cri issue on the containerd server, feel free to add your init to the cri init.. we can always move this out of the cri plugin and into it's own plugin down the road if need

on the run once thing something like:
```
var rdtEnabled bool
var rdtEnabledOnce sync.Once

func GetRdtEnabled() {
   return rdtEnabled
}

then in the init.. 
    var result Error
    rdtEnabledOnce.Do(func() {
	rdtEnabled = true // if passes all the enabled checks
        // should be false and set the result error that cause the fail..
	})
       return (some go func for monitoring this capability?), result 
```

If you don't move your init to the cri plugin, and a monitoring thread is not needed for this plugin, I guess it's ok to just stub it? @dmcgowan thoughts?

marquiz replied:
Well, it's not CRI only. It is possible to set the RDT class with `ctr`, too. E.g. something like
```
ctr run --rdt-class my-class ...
```

samuelkarp commented on pkg/rdt/rdt.go:
```diff
@@ -0,0 +1,90 @@
+import (
+	"fmt"
+
+	"github.com/containerd/containerd/log"
+	"github.com/intel/goresctrl/pkg/rdt"
+	"github.com/pkg/errors"
+
+	"github.com/containerd/containerd/plugin"
```
nit: these are grouped weirdly.  Can you do something like this instead?

```suggestion
	"github.com/containerd/containerd/log"
	"github.com/containerd/containerd/plugin"

	"github.com/intel/goresctrl/pkg/rdt"
	"github.com/pkg/errors"
```

This keeps the imports that are part of this module together and the imports coming from other modules separate.

marquiz replied:
Indeed, dunno why it was grouped that way 🧐  Now fixed

marquiz commented:
Rebased.

One change in the first patch: added `oci.WithRdt` stub for other platforms than linux.

marquiz commented:
Rebased

mikebrow commented on pkg/cri/server/container_create_linux.go:
```diff
@@ -260,6 +261,15 @@ func (c *criService) containerSpec(
 
 	supplementalGroups := securityContext.GetSupplementalGroups()
 
+	// Get RDT class
```
do we need a config flag to ignoreRdtNotEnabledErrors default false..

mikebrow replied:
if you go the cri inited and configured route ^ the ignore flag and config path string can be added to the cri config as well

marquiz replied:
I added a patch introducing `cri.containerd.ignore_rdt_not_enabled_errors`config flag. I felt cri plugin was the appropriate place for this config flag. Maybe this patch should be split into a separate PR(?)

mikebrow replied:
one is fine.. because it's related imo.. (note maintainers don't have squash merge on this repo.., someone might ask for the commits to be squashed around/after review approvals)

marquiz replied:
> one is fine..because it's related imo..

👍  I was pondering a separate PR for the case this new patch creates a lot of debate. I.e. basically removing extra friction by splitting it out. But let's keep it here, for now

> (note maintainers don't have squash merge on this repo.., someone might ask for the commits to be squashed around/after review approvals)

Sure, I can do that if asked. My personal style is just splitting changes into discrete, logically and functionally self-contained commits. In the hope of helping understand the in-code dependencies, making it easier to read for reviewers and also future readers (who are looking at the commit history)

mxpv replied:
Should we just ignore RDT if its not supported/enabled? `IgnoreRdtNotEnabledErrors` feels hacky.

marquiz replied:
> RDT is for cache and memory bandwidth management. Is it worth failing?

I guess it really depends on the usage scenario. There are use cases with high SLA/QoS requirements where you certainly want to fail if it fails. I agree that `IgnoreRdtNotEnabledErrors` sounds hackish, maybe changing the name could ease the pain?

mxpv replied:
`IgnoreRdtErrors` may be?

marquiz replied:
`IgnoreRdtErrors` would be semantically slightly different, i.e. I think with this we should ignore all rdt-related errors (that happen in the cri plugin). Whereas `IgnoreRdtNotEnabledErrors` now ignores errors only in the case that rdt has not been enabled (in containerd scope) – i.e. errors are not ignored if rdt has been (successfully) enabled by the sysadmin. I'm basically fine with either one.

In practice, this is related to the annotation parsing that happens now on the runtime side. Going forward, my goal is to get rdt in the cri protocol, after which rdt prolly changes into a "pass-through mode" where the client sets the rdt class in the containercreate request and the cri plugin just passes this into the oci runtime spec. Errors will happen in runc side.

mikebrow commented on pkg/cri/server/container_create_linux.go:
```diff
@@ -260,6 +261,15 @@ func (c *criService) containerSpec(
+	// Get RDT class
+	rdtClass, err := rdt.ContainerClassFromAnnotations(config.GetMetadata().GetName(), config.Annotations, sandboxConfig.Annotations)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid RDT class")
```
two types of errors here.. invalid class and not enabled.. needs a generic string on the wrap

marquiz replied:
Changed wording to `failed to set RDT class` which covers all types of errors

mikebrow reviewed (commented):
see comments

marquiz commented:
> see comments

Replied/responded and rebased. Thanks for the review @mikebrow!

Any feedback from you @samuelkarp @mxpv?

marquiz commented:
Rebased.

woof @mxpv @mikebrow

mikebrow commented on pkg/cri/server/container_create_linux.go:
```diff
@@ -260,7 +261,12 @@ func (c *criService) containerSpec(
-		return nil, errors.Wrap(err, "failed to set RDT class")
+		if !rdt.Enabled() && c.config.ContainerdConfig.IgnoreRdtNotEnabledErrors {
+			rdtClass = ""
+			logrus.Info(err)
```
how about logrus.Debugf("continuing create container %s, ignoring rdt not enabled %v", containerName, err)

marquiz replied:
changed to
```
logrus.Debugf("continuing create container %s, ignoring rdt not enabled (%v)", containerName, err)
```

marquiz commented:
Rebased and integration point in the containerd codebase changed.

As per discussions with @mikebrow I now scrapped the separate `rdt` internal plugin. Instead, I hooked it into the `tasks-service` plugin. One collateral of this (the way this is currently implemented) is that an error in RDT configuration (invalid/non-existent config file, for example) also causes the `tasks-service` plugin initialization to fail 🧐 

Ping @mxpv @mikebrow @dmcgowan @AkihiroSuda

mxpv commented:
We don't want task service plugin to fail, why not a separate plugin?
For instance devmapper plugin fails by default in containerd:

```
...
loading plugin "io.containerd.metadata.v1.bolt"...  type=io.containerd.metadata.v1
could not use snapshotter devmapper in metadata plugin  error="devmapper not configured"
...
```

Can we do same for rdt?

marquiz commented:
Thanks for the review @mxpv 

> We don't want task service plugin to fail

Yeah, I started to think that myself, doesn't really make any sense 🙄 I changed it now to just an error msg in the log.

> why not a separate plugin?

Dunno, fear of excessive number of plugins / abuse of the plugin mechanism(?) 🤷  That was the guidance I got 😸 

> For instance devmapper plugin fails by default in containerd:

RDT now behaves in a similar manner (except it doesn't fail by default as the config file path is empty). There is an error message in the log (and RDT features cannot be used) but otherwise life's good.

mxpv commented:
> That was the guidance I got 😸

was there a discussion I missed? A plugin seems ok to me.

marquiz commented:
> > That was the guidance I got 😸
> 
> was there a discussion I missed? A plugin seems ok to me.

Maybe @mikebrow can elaborate on this. Got that feedback through some slack discussions.

Anyway, I can change this  back to a separate plugin if others are fine with it 😅 I'm pretty indifferent on the point of integration, whatever goes for me.

mxpv reviewed (approved):
Ok, changes look good, don't want to block on plugin thing. We can figure it out later.

marquiz commented:
> Ok, changes look good, don't want to block on plugin thing. We can figure it out later.

Cool, as said I'm pretty flexible with any guidance and can easily change things 😎

mikebrow commented:
> Rebased and integration point in the containerd codebase changed.
> 
> As per discussions with @mikebrow I now scrapped the separate `rdt` internal plugin. Instead, I hooked it into the `tasks-service` plugin. One collateral of this (the way this is currently implemented) is that an error in RDT configuration (invalid/non-existent config file, for example) also causes the `tasks-service` plugin initialization to fail 🧐
> 
> Ping @mxpv @mikebrow @dmcgowan @AkihiroSuda

there is no reason task should fail if rdt init fails

mikebrow commented:
> > That was the guidance I got 😸
> 
> was there a discussion I missed? A plugin seems ok to me.

framing the discussion ... Should we add new plugins rdt, blockio, cdi.. maybe others... for making edits to the oci spec of new containers.  If we want to switch to using plugins for these spec edits, does containerd/NRI make more sense than adding new core services?  If not NRI how will these plugins interact with cri/ctr/nerdctl/... do we add them as dependencies to each client? If not core service default, or nri plugins...should cri check if they are installed, then status?

marquiz commented:
> for making edits to the oci spec of new containers.

I'd argue that especially in the case of rdt, it (the "plugin") is not that much about "oci spec edits" but merely about enabling/configuring rdt and being aware what classes are available.

mikebrow commented:
> > for making edits to the oci spec of new containers.
> 
> I'd argue that especially in the case of rdt, it (the "plugin") is not that much about "oci spec edits" but merely about enabling/configuring rdt and being aware what classes are available.

```
// WithRdt sets the container's RDT parameters
 func WithRdt(closID, l3CacheSchema, memBwSchema string) SpecOpts {
 	return func(ctx context.Context, _ Client, c *containers.Container, s *Spec) error {
 		s.Linux.IntelRdt = &specs.LinuxIntelRdt{
 			ClosID:        closID,
 			L3CacheSchema: l3CacheSchema,
 			MemBwSchema:   memBwSchema,
 		}
 		return nil
 	}
 }
```

did not mean to place a negative connotation on any of the desired/meaningful edits to the OCI spec of new containers.. to enable/configure such features

marquiz commented:
Rebased.

Ping @mikebrow

mikebrow reviewed (approved):
LGTM
see two nits..

marquiz commented on pkg/cri/config/config.go:
```diff
@@ -97,6 +97,10 @@ type ContainerdConfig struct {
 	// layers to the snapshotter.
 	DiscardUnpackedLayers bool `toml:"discard_unpacked_layers" json:"discardUnpackedLayers"`
+
+	// IgnoreRdtNotEnabledErrors is a boolean flag flag ignore RDT related
```
Thanks for spotting this. Fixed

marquiz commented on cmd/ctr/commands/commands.go:
```diff
@@ -185,6 +185,10 @@ var (
 		},
+		cli.StringFlag{
+			Name:  "rdt-class",
+			Usage: "name of the RDT class (or CLOS) to associate the container with",
```
Changed as suggested

marquiz commented:
Addressed latest feedback.

@mikebrow PTAL

mikebrow reviewed (approved):
LGTM

marquiz commented:
Annybody able to re-trigger the failing tests? Or do we care?

dmcgowan commented:
We do re-run the tests until green. We fixed the outstanding CI flakiness from last week. Can you rebase this one and we can get it green

marquiz commented:
Thanks @dmcgowan! Rebased

> We do re-run the tests until green. We fixed the outstanding CI flakiness from last week. Can you rebase this one and we can get it green

Cool 🤞

marquiz commented:
ping @dmcgowan

askervin commented:
Now that this patch is approved, I have rewritten #5490 to follow exactly the pattern introduced in this patch.

#5490 is developed on top of this patch to avoid conflicts, as both patches change the same workflow and files. Please merge this patch first.

marquiz commented:
Rebased. ping @dmcgowan
